README.rst
Memo
====

Online memorization tool optimizing your long term memorization and boosting your productivity comfortably in the background.

.. image:: https://img.shields.io/badge/built%20with-Cookiecutter%20Django-ff69b4.svg
     :target: https://github.com/pydanny/cookiecutter-django/
     :alt: Built with Cookiecutter Django
.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
     :target: https://github.com/ambv/black
     :alt: Black code style


Settings
--------

Moved to settings_.

.. _settings: http://cookiecutter-django.readthedocs.io/en/latest/settings.html

Basic Commands
--------------

Setting Up Your Users
^^^^^^^^^^^^^^^^^^^^^

* To create a **normal user account**, just go to Sign Up and fill out the form. Once you submit it, you'll see a "Verify Your E-mail Address" page. Go to your console to see a simulated email verification message. Copy the link into your browser. Now the user's email should be verified and ready to go.

* To create an **superuser account**, use this command::

    $ python manage.py createsuperuser

For convenience, you can keep your normal user logged in on Chrome and your superuser logged in on Firefox (or similar), so that you can see how the site behaves for both kinds of users.

Type checks
^^^^^^^^^^^

Running type checks with mypy:

::

  $ mypy memo

Test coverage
^^^^^^^^^^^^^

To run the tests, check your test coverage, and generate an HTML coverage report::

    $ coverage run -m pytest
    $ coverage html
    $ open htmlcov/index.html

Running tests with py.test
~~~~~~~~~~~~~~~~~~~~~~~~~~

::

  $ pytest

Live reloading and Sass CSS compilation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Moved to `Live reloading and SASS compilation`_.

.. _`Live reloading and SASS compilation`: http://cookiecutter-django.readthedocs.io/en/latest/live-reloading-and-sass-compilation.html



Celery
^^^^^^

This app comes with Celery.

To run a celery worker:

.. code-block:: bash

    cd memo
    celery -A config.celery_app worker -l info

Please note: For Celery's import magic to work, it is important *where* the celery commands are run. If you are in the same folder with *manage.py*, you should be right.




Email Server
^^^^^^^^^^^^

In development, it is often nice to be able to see emails that are being sent from your application. If you choose to use `MailHog`_ when generating the project a local SMTP server with a web interface will be available.

#. `Download the latest MailHog release`_ for your OS.

#. Rename the build to ``MailHog``.

#. Copy the file to the project root.

#. Make it executable: ::

    $ chmod +x MailHog

#. Spin up another terminal window and start it there: ::

    ./MailHog

#. Check out `<http://127.0.0.1:8025/>`_ to see how it goes.

Now you have your own mail server running locally, ready to receive whatever you send it.

.. _`Download the latest MailHog release`: https://github.com/mailhog/MailHog/releases

.. _mailhog: https://github.com/mailhog/MailHog



Sentry
^^^^^^

Sentry is an error logging aggregator service. You can sign up for a free account at  https://sentry.io/signup/?code=cookiecutter  or download and host it yourself.
The system is setup with reasonable defaults, including 404 logging and integration with the WSGI application.

You must set the DSN url in production.


Deployment
----------

The following details how to deploy this application.


Heroku
^^^^^^

See detailed `cookiecutter-django Heroku documentation`_.

.. _`cookiecutter-django Heroku documentation`: http://cookiecutter-django.readthedocs.io/en/latest/deployment-on-heroku.html






CHANGELOG
== Version 0.0.2.3 ==
OK cards max height 300px and card_body should be scrollable
OK Remove defunct cardset app
OK migrate to eu region
OK !!! BUGFIX Join Group
    https://www.braingain.ai/group/welcome-to-42-wolfsburg-c-lang-bootcamp
    -> 500 error
OK BUGFIX: Edit Group on GroupDetail not working
OK Make crude bookkeeping for the project (in/out per month)
OK Make "42 Wolfsburg C Lang Bootcamp" Studygroup with a set of C cards on production (keywords & norms)
OK Add menues admin directory and handle empty directory
OK OP Get first users...
OK Unobstuctive changes to Interface
OK Make delete card only for admin !!!
OK Disable create group buttons

== Version 0.0.2.2 == Membership Rules
OK Make all forms look bootstrapy and nice (crispy form)
OK BUGFIX: CKEDITOR slows down page
OK Add Privacy & data protection from friendcheck and link in footer
OK Make external resources local
OK Optimize Membership and Rules
    template tag "get_group_permission(user,group)
OK   BUGFIX Joining group fails silently
OK Add Google Analytics to the mix

== Version 0.0.2.1 == Redo Recall/Train => GainView
OK GainView
    OK FormView with options:
        - mode: train/recall
        - group: user groups
        - topic: group topic or all
    OK View template
    OK test function
    OK remove old code
OK BUGFIX: On user delete the membership get's deleted but NOT the main user space group. The main user group should be deleted and its content.
OK Add to CKEditor: basics styling abilities
OK Allow edit card from train view
OK Make messages fade away
OK Design/Improve Frontend with HTML/CSS Bootstrap 4
    OK Add slim Footer with important links
    OK Group card and card count and member count for group
    OK Icons for Group, Card and Topic and Manage Button

== Version 0.0.1.1 == Optimize GroupDetailView
OK Redirect LogIn to Group List
OK Make Frontpage describing braingain

== Version 0.0.1 == "Everything works!"
OK Connect braingain.ai
OK Add Edit Settings modal on card
OK Display recall score, is paused and priority on card (bottom-left)
OK Invite System: link to join a studygroup => login if not auth
    OK Show number of members and cards on group card
    OK Group Card: show invite link to the group, for disemmanation
    OK Check functionality
OK Enable https
OK connect mailgun for app
OK Test Permissions work
OK what happens after fresh signup or login
OK mailgun?

== Version 0.0.0 ==
OK Edit & Create Group should go in modal (linked from Manage Group)
OK Make paginate appear only below the list and only when num_pages > 1
OK Group DetailView
    OK   paginate
    OK   search card by text / topic
OK Design the train and recall templates to resemble cards
OK test/train does not get all cards
OK Fetch least learned and recalled
OK Make Footer with Copyright (take from example)

== Alpha == (Test App)
OK Design model structure
OK REFACTOR Code (Model names and derived identifiers...)
OK Connect studygroup view with cardsets
OK Implement Studygroups (studygroups) with permission, invite system & ranking system.
    OK Initialize app and model
    OK Add managing views for frontend
        OK create OK edit OK delete OK join OK unjoin OK view
OK Lifecycle Analysis & Implementation
    OK Group:
        OK Implement Group Permissions
            OK can_create_group  => is_auth(user) & has_free_group_slot(user)
            OK can_view_group    => is_auth(user) & is_group_member(user, group)
            OK can_update_group  => is_auth(user) & is_group_admin(user, group)
            OK can_delete_group  => is_auth(user) & is_group_admin(user, group)
            OK can_join_group    => user.is_auth & user.is_not_group_member(user, group) & user.has_free_group_slot
            OK can_leave_group   => user.is_auth & user.is_group_member(user, group) & not admin
            OK can_manage_member => is_auth(user) & is_group_admin(user, group)
            OK can_add_card      => is_auth(user) & is_group_editor(user, group) | is_group_admin(user, group)
            OK can_add_topic     => is_auth(user) & is_group_editor(user, group) | is_group_admin(user, group)
        OK Implement Group Lifecycle Views
            OK CreateView
            OK DetailView
            OK UpdateView
            OK DeleteView
            OK JoinView
            OK LeaveView
            OK ManageMembershipView
        OK Implement Group Lifecycle Templates
    OK   Topic:
        OK Implement Topic Permissions
        OK Implement Topic Lifecycle Views
            OK CreateView
            OK DetailView
            OK UpdateView
            OK DeleteView
        OK Implement Topic Lifecycle Templates
    OK Card:
        OK Implement Card Permissions
        OK Implement Card Lifecycle Views
            OK CreateView
            OK DetailView
            OK UpdateView
            OK DeleteView
        OK Implement Card Lifecycle Templates
            OK Make partial _card_card.html toogle front_text & back_text
    OK   Performance:
        OK Implement Performance Permissions
        OK Implement Performance Lifecycle Views
            train
            test
        OK Implement Performance Lifecycle Templates

== Proof of Concept ==
OK Add Basic Workflow MemoSet CRUD to frontend
OK General Main Brain on user.create signal
OK Add Logo
OK Add Basic Workflow MemoCard CRUD to frontend
OK Fix accordion collapse and style errors in memoset_root_list_view
OK Make template tag for getting the performance model based on the request user and a memocard
OK Implement Basic Gain Workflow
    OK Training (cycle cards)
    OK Recall (test cards)
OK Connect sentry
OK Add fields to CardPerformance {
        OK unique_id:
        OK priority: (low,normal,high) (sort in qs)
        OK is_paused: False (exclude from qs)}
    OK Redo Queryset methods
OK View for update PerformanceSettings
    OK Add prio and is_paused to card edit to card performance edit
OK BUGFIX:
    OK /memo/manage/set/root/list => rename in /memo/manage/brains and cleanup html

config/urls.py
from django.conf import settings
from django.conf.urls.static import static
from django.contrib import admin
from django.urls import include, path, reverse_lazy
from django.views import defaults as default_views
from django.views.generic import RedirectView, TemplateView
from rest_framework.authtoken.views import obtain_auth_token


class HomeRedirectView(RedirectView):
    def get_redirect_url(self, *args, **kwargs):
        # redirects authenticated users to their brain list
        if self.request.user.is_authenticated:
            return reverse_lazy("studygroups:group_list_view")
        return reverse_lazy("start")


urlpatterns = [
    path("", HomeRedirectView.as_view(), name="home"),
    path("start/", TemplateView.as_view(template_name="pages/home.html"), name="start"),
    path(
        "about/", TemplateView.as_view(template_name="pages/about.html"), name="about"
    ),
    path(
        "about/",
        TemplateView.as_view(template_name="pages/about.html"),
        name="about",
    ),
    path(
        "terms_and_conditions/",
        TemplateView.as_view(template_name="pages/terms_and_conditions.html"),
        name="terms_and_conditions",
    ),
    path(
        "privacy_policy/",
        TemplateView.as_view(template_name="pages/privacy_policy.html"),
        name="privacy_policy",
    ),
    # Django Admin, use {% url 'admin:index' %}
    path(settings.ADMIN_URL, admin.site.urls),
    # User management
    path("users/", include("memo.users.urls", namespace="users")),
    path("accounts/", include("allauth.urls")),
    #
    # App url confs
    #
    path("group/", include("memo.studygroups.urls", namespace="studygroup")),
    path("card/", include("memo.flashcards.urls", namespace="flashcard")),
] + static(settings.MEDIA_URL, document_root=settings.MEDIA_ROOT)

# API URLS
urlpatterns += [
    # API base url
    path("api/", include("config.api_router")),
    # DRF auth token
    path("auth-token/", obtain_auth_token),
]

if settings.DEBUG:
    # This allows the error pages to be debugged during development, just visit
    # these url in browser to see how these error pages look like.
    urlpatterns += [
        path(
            "400/",
            default_views.bad_request,
            kwargs={"exception": Exception("Bad Request!")},
        ),
        path(
            "403/",
            default_views.permission_denied,
            kwargs={"exception": Exception("Permission Denied")},
        ),
        path(
            "404/",
            default_views.page_not_found,
            kwargs={"exception": Exception("Page not Found")},
        ),
        path("500/", default_views.server_error),
    ]
    if "debug_toolbar" in settings.INSTALLED_APPS:
        import debug_toolbar

        urlpatterns = [path("__debug__/", include(debug_toolbar.urls))] + urlpatterns
